File: sdk/go/client/groups.go
package client

import (
	"context"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
)

// GroupsService 封装 Client API 的 Groups 服务。
type GroupsService struct{ c *Client }

// CreateGroup 创建用户组。
func (t *GroupsService) CreateGroup(ctx context.Context, name string) (*clientv1.Group, error) {
	return t.c.groups.CreateGroup(ctx, &clientv1.CreateGroupRequest{Name: name})
}

// GetGroup 按 ID 获取用户组。
func (t *GroupsService) GetGroup(ctx context.Context, groupID string) (*clientv1.Group, error) {
	return t.c.groups.GetGroup(ctx, &clientv1.GetGroupRequest{Id: groupID})
}

// ListGroups 列出当前用户所属用户组。
func (t *GroupsService) ListGroups(ctx context.Context) (*clientv1.ListGroupsResponse, error) {
	return t.c.groups.ListGroups(ctx, &sharedv1.ListRequest{})
}

// CreateMembership 邀请成员加入用户组（按邮箱）。
func (t *GroupsService) CreateMembership(ctx context.Context, groupID, email, name string, roles []string) (*clientv1.Membership, error) {
	return t.c.groups.CreateMembership(ctx, &clientv1.CreateMembershipRequest{
		GroupId: groupID,
		Email:   email,
		Name:    name,
		Roles:   roles,
	})
}

// ListMemberships 列出用户组成员。
func (t *GroupsService) ListMemberships(ctx context.Context, groupID string) (*clientv1.ListMembershipsResponse, error) {
	return t.c.groups.ListMemberships(ctx, &clientv1.ListMembershipsRequest{GroupId: groupID})
}

// UpdateMembershipStatus 更新成员状态（如 active / invited / banned）。
func (t *GroupsService) UpdateMembershipStatus(ctx context.Context, groupID, membershipID, status string) (*clientv1.Membership, error) {
	return t.c.groups.UpdateMembershipStatus(ctx, &clientv1.UpdateMembershipStatusRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
		Status:       status,
	})
}

// DeleteGroup 删除用户组（仅 owner 角色；Round3 H4-3 补齐与 TS/proto 对齐）。
func (t *GroupsService) DeleteGroup(ctx context.Context, groupID string) error {
	_, err := t.c.groups.DeleteGroup(ctx, &clientv1.GetGroupRequest{Id: groupID})
	return err
}

// DeleteMembership 移除用户组成员。
func (t *GroupsService) DeleteMembership(ctx context.Context, groupID, membershipID string) error {
	_, err := t.c.groups.DeleteMembership(ctx, &clientv1.GetMembershipRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
	})
	return err
}

File: sdk/typescript/src/client/groups.ts
import { listQuery, type HttpTransport } from "../http.js";
import type { ListParams, Membership, Group } from "../types.js";

export class ClientGroupsService {
  constructor(private readonly http: HttpTransport) {}

  async createGroup(name: string): Promise<Group> {
    return this.http.request<Group>("POST", "/v1/groups", { body: { name } });
  }

  async listGroups(params?: ListParams): Promise<Group[]> {
    const res = await this.http.request<{ groups: Group[] }>("GET", "/v1/groups", {
      query: listQuery(params),
    });
    return res.groups ?? [];
  }

  async getGroup(id: string): Promise<Group> {
    return this.http.request<Group>("GET", `/v1/groups/${id}`);
  }

  async deleteGroup(id: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/groups/${id}`);
  }

  async createMembership(
    groupId: string,
    input: { email: string; name?: string; roles?: string[] }
  ): Promise<Membership> {
    return this.http.request<Membership>("POST", `/v1/groups/${groupId}/memberships`, {
      body: {
        group_id: groupId,
        email: input.email,
        name: input.name ?? "",
        roles: input.roles,
      },
    });
  }

  async listMemberships(groupId: string): Promise<Membership[]> {
    const res = await this.http.request<{ memberships: Membership[] }>(
      "GET",
      `/v1/groups/${groupId}/memberships`
    );
    return res.memberships ?? [];
  }

  async updateMembershipStatus(
    groupId: string,
    membershipId: string,
    status: "accepted" | "rejected"
  ): Promise<Membership> {
    return this.http.request<Membership>(
      "PATCH",
      `/v1/groups/${groupId}/memberships/${membershipId}/status`,
      {
        body: {
          group_id: groupId,
          membership_id: membershipId,
          status,
        },
      }
    );
  }

  async deleteMembership(groupId: string, membershipId: string): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/groups/${groupId}/memberships/${membershipId}`
    );
  }
}
